File: src/hittable/hittablelist.cpp
#include "hittablelist.hpp"

auto HittableList::hit(const Ray &ray, const f64 t_min, const f64 t_max) const -> std::optional<HitRecord> {
    std::optional<HitRecord> rec;
    f64 closest = t_max;

    for (const auto &hittable: hittables)
        if (const auto subrec = hittable->hit(ray, t_min, closest); subrec.has_value())
            closest = subrec.value().t, rec = subrec;

    return rec;
}

auto HittableList::bounding_box(const f64 t0, const f64 t1) const -> std::optional<AABB> {
    if (hittables.empty()) return std::optional<AABB>();
    std::optional<AABB> box;

    for (const auto &hittable: hittables)
        if (const auto subbox = hittable->bounding_box(t0, t1); subbox.has_value()) {
            if (box.has_value())
                box.emplace(AABB::surrounding_box(box.value(), subbox.value()));
            else box = subbox;
        } else return std::optional<AABB>();

    return box;
}

File: src/hittable/hittablelist.rs
use crate::ray::Ray;
use crate::hittable::*;

#[derive(Clone, Default)]
pub struct HittableList {
    objects: Vec<Rc<dyn Hittable>>,
}

impl HittableList {
    pub fn clear(&mut self) {
        self.objects.clear();
    }

    pub fn push<T: Hittable + 'static>(&mut self, object: T) {
        self.objects.push(Rc::new(object));
    }
    pub fn push_ptr<T: Hittable + 'static>(&mut self, ptr: Rc<T>) {
        self.objects.push(ptr);
    }
}

impl Hittable for HittableList {
    fn hit(&self, ray: &Ray, t_min: f64, t_max: f64) -> Option<HitRecord> {
        let mut sol = None;
        let mut closest = t_max;

        for object in &self.objects {
            if let Some(subsol) = object.hit(&ray, t_min, closest) {
                closest = subsol.t;
                sol = Some(subsol);
            }
        }

        sol
    }
}
